include/nbl/video/IGPUObjectFromAssetConverter.h

```diff
@@ -591,76 +710,189 @@ auto IGPUObjectFromAssetConverter::create(const asset::ICPUMesh** const _begin,
 		auto count = regions.size();
 		if (count)
 		{
-            // TODO: @Criss why isn't this buffer caches and why are we not going through recursive asset creation and getting ICPUBuffer equivalents? (we can always discard/not cache the GPU Buffers created only for image data upload)
-			auto tmpBuff = m_driver->createFilledDeviceLocalGPUBufferOnDedMem(cpuimg->getBuffer()->getSize(),cpuimg->getBuffer()->getPointer());
-			m_driver->copyBufferToImage(tmpBuff.get(),gpuimg.get(),count,regions.begin());
-            if (!integerFmt)
             {
-                uint32_t lowestPresentMip = 1u;
-                while (cpuimg->getRegions(lowestPresentMip).size())
-                    lowestPresentMip++;
-                // generate temporary image view to make sure we don't screw up any explicit mip levels
-                IGPUImageView::SCreationParams tmpViewParams;
-                tmpViewParams.subresourceRange.levelCount = params.mipLevels+1u-lowestPresentMip;
-                // if not all mip levels have been manually specified
-                if (tmpViewParams.subresourceRange.levelCount>1u)
-                {
-                    tmpViewParams.flags = static_cast<IGPUImageView::E_CREATE_FLAGS>(0u);
-                    tmpViewParams.image = core::smart_refctd_ptr(gpuimg);
-                    switch (params.type)
-                    {
-                        case asset::IImage::ET_1D:
-                            tmpViewParams.viewType = IGPUImageView::ET_1D_ARRAY;
-                            break;
-                        case asset::IImage::ET_2D:
-                            if (params.flags & asset::IImage::ECF_CUBE_COMPATIBLE_BIT)
-                              tmpViewParams.viewType = IGPUImageView::ET_CUBE_MAP_ARRAY;
-                            else
-                              tmpViewParams.viewType = IGPUImageView::ET_2D_ARRAY;
-                            break;
-                        case asset::IImage::ET_3D:
-                            tmpViewParams.viewType = IGPUImageView::ET_3D;
-                            break;
-                        default:
-                            assert(false);
-                            break;
-                    }
-                    tmpViewParams.format = params.format;
-                    //tmpViewParams.subresourceRange.aspectMask
-                    tmpViewParams.subresourceRange.baseMipLevel = lowestPresentMip-1u;
-                    tmpViewParams.subresourceRange.layerCount = params.arrayLayers;
-                    auto tmpView = m_driver->createGPUImageView(std::move(tmpViewParams));
-                    // deprecated OpenGL path (do with compute shader in the future)
-                    tmpView->regenerateMipMapLevels();
-                }
+                IGPUCommandBuffer::SImageMemoryBarrier imgbarrier;
+                // TODO layouts
+                imgbarrier.oldLayout = asset::EIL_UNDEFINED;
+                imgbarrier.newLayout = asset::EIL_UNDEFINED;
```

Comment:
old layout is TRANSFER iirc
new layout should be optimal for shader reading or GENERAL (should be an option tbh, because GENERAL allows all usages, iirc, but has a perf penalty compared to shader read which is 99% the case for textures made with the auto converter)

Reply:
when creating images, take the user supplied image usage bits, but then force:
- transfer bit, if there are regions to transfer to the image
- the bits for the compute mipmapper, if there are mipmaps to compute

VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL should be the initial layout

Then from the final usage bits deduce if you need to transfer to a layout:
- VK_IMAGE_LAYOUT_GENERAL (if mipmaps need to be made or we have usage bits indicating writing)
- VK_IMAGE_LAYOUT_READ_ONLY_OPTIMAL (no mipmaps need computing, but there are some usage bit which preclude)
- VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL_KHR (usage is only sampling)